site/docs/plugins/runner.md

```diff
@@ -78,10 +80,299 @@ run(bot);
+
+We are now going to look at more advanced usage of the plugin.
+
+The supplied constraint function can be used not only to specify chat identifier, or user identifier.
+Instead, you can return _a list of constraint identifier strings_ that determine for every update individually what other computations it must wait for before processing can begin.
+
+As an example, you could return both the chat identifier, and the user identifier of the message author.
+
+```ts
+bot.use(sequentialize((ctx) => {
+  const chat = ctx.chat?.id.toString();
+  const user = ctx.from?.id.toString();
+  return [chat, user].filter((con) => con !== undefined);
+}));
+```
+
+This would make sure that messages in the same chat are ordered correctly.
+In addition, if Alice sends message in a group, and then sends a message to your bot in the private chat, then these two messages are ordered correctly.
+
+In a sense, you can therefore specify a graph of dependencies between updates.
+grammY runner will resolve all necessary constraints on the fly and block those updates as long as necessary to ensure correct message ordering.
+
+The implementation of this is very efficient.
+It needs constant memory (unless you specify infinite concurrency), and it needs (amortized) constant processing time per update.
+
+## Graceful Shutdown
+
+In order for the bot to complete its work correctly you [should signal](../advanced/reliability.md#using-grammy-runner) it to stop when the process is about to be destroyed.
+
+Note that you can wait for the runner to terminate by `await`ing the `task` in the [RunnerHandle](https://deno.land/x/grammy_runner/mod.ts?s=RunnerHandle) returned from `run`.
+
+```ts
+const handle = run(bot);
+
+handle.task.then(() => {
+  console.log("Bot done processing!");
+});
+```
+
+## Advanced Options
+
+grammY runner consists of three things: a source, a runner, and a sink.
+The source pulls in updates, the sink consumes updates, and the runner configures and connects the two.
+
+> An in-depth description on how the runner workds internally can be found [down here](#how-it-works-behind-the-scenes).
+
+Each of these three parts can be configured through various options.
+This can reduce the network traffic, let you specify allowed updates, and more.
+
+Each part of the runner accepts its configuration through a dedicated options object.
+
+```ts
+run(bot, {
+  source: {},
+  runner: {},
+  sink: {},
+});
+```
+
+You should check out the `RunOptions` in the [API reference](https://deno.land/x/grammy_runner/mod.ts?s=RunOptions) to see which options are available.
+
+For example, you will find there that `allowed_updates` can be enabled using the following code snippet.
+
+```ts
+run(bot, { runner: { fetch: { allowed_updates: [] } } });
+```
+
+## Multithreading
+
+JavaScript is single-threaded.
+This is amazing because [concurrency is hard](../advanced/scaling.md#concurrency-is-hard) so if there is only a single thread, then this removes a lot of headache.
```

Comment:
```suggestion
This is amazing because [concurrency is hard](../advanced/scaling.md#concurrency-is-hard), meaning that if there is only a single thread, a lot of headache is automatically removed.
```

Reply:
Is automatically removed? Sounds like some technical automation

Reply:
```suggestion
This is amazing because [concurrency is hard](../advanced/scaling.md#concurrency-is-hard), meaning that if there is only a single thread, a lot of headache is naturally removed.
```